### ISSUE TITLE
k0s token create uses wrong ip address

### ISSUE THREAD

**Issue by mmack** [k0sproject/k0s#704]
**Version**
```
v0.10.0
```
**Platform**
Which platform did you run k0s on?
```
Ubuntu 20.04.1 LTS
```
**What happened?**
Worker tries to contact the master API on the wrong IP Address.

**How To Reproduce**
k0s server config:
```
apiVersion: k0s.k0sproject.io/v1beta1
kind: Cluster
metadata:
  name: k0s
spec:
  api:
    address: 10.1.1.1
    externalAddress: 10.1.1.1
    sans:
    - 10.1.1.1
    extraArgs:
      feature-gates: "RemoveSelfLink=false" # needed for metallb
  controllerManager:
    extraArgs: {}
  scheduler:
    extraArgs: {}
  storage:
    type: etcd
    etcd:
      peerAddress: 10.1.1.1
  network:
    podCIDR: "10.244.0.0/16"
    serviceCIDR: "10.96.0.0/12"
    calico:
      mode: "bird"
    dualStack:
      enabled: true
      IPv6podCIDR: "fd00::/108"
      IPv6serviceCIDR: "fd01::/108"
```
10.1.1.1: The ip of a wireguard interface.

`k0s token create -c /etc/k0s/k0s.yaml --role=worker | base64 -d | gunzip -`:
```
apiVersion: v1
clusters:
- cluster:
    server: https://X.X.X:X:6443
```
X.X.X.X is the external ip of the default interface. This should be 10.1.1.1 imho.

**Expected behavior**
If i set a externalAddress / address in the k0s server config manually i expect the `k0s token create` command to use this ip instead of using anything else.

And btw: I think the same happens on generating a kubeconf.

So to sum this up: If ip's are "detected" the behaviour is fine. If they are set manually i think it's not.

**Comment by mmack** [comment]
I digged a little further and it seems if if i do not set "externalAddress", the generated tokens use 10.1.1.1 as API server...

**Comment by ruskofd** [comment]
Same here :

```
kind: Cluster
metadata:
  name: k0s
spec:
  api:
    address: 10.0.20.11
  controllerManager:
    extraArgs: {}
  scheduler:
    extraArgs: {}
  storage:
    type: etcd
    etcd:
      peerAddress: 10.0.20.11
  ```
  
  It seems it takes the primary IP of my machine. I don't use external address for now (Ansible testing)
```
  apiVersion: v1
clusters:
- cluster:
    server: https://192.168.122.200:9443
  ```
 
I tried with a 5 controllers and 3 workers cluster, on all machines there is the same issue.

  **Version**
```
v0.10.0
```

**Platform**
```
Fedora 33
```

**Comment by jnummelin** [comment]
I cannot reproduce this. 🤔 

With config:
```
kind: Cluster
metadata:
  name: k0s
spec:
  api:
    address: 10.0.20.11
    externalAddress: foo.bar.com
  
```
I get:
```
/ # ./k0s token create -c k0s.yaml --role=worker | base64 -d | gunzip - | grep "server:"
    server: https://foo.bar.com:6443
```

With config:
```
kind: Cluster
metadata:
  name: k0s
spec:
  api:
    address: 10.0.20.11
```

I get:
```
/ # ./k0s token create -c k0s.yaml --role=worker | base64 -d | gunzip - | grep "server:"
    server: https://10.0.20.11:6443
```

So the "order of preference" for the address to be used in tokens is:
- `api.externalAddress`; This is usually LB address when running many controllers
- `api.address`: The "external" IP of controller node
- "autodetect"; Basically gets the first non-loopback address it finds

@mmack @ruskofd So few things to check out:
- The `k0s token create -c k0s.yaml` is pointing to the same config as the server is using
- The config file is really readable by the user executing the token command
- `k0s validate config /path/to/k0s.yaml` is able to validate the config file

There is a bug that if the token command fails to parse the config, it'll just silently revert to use default config, in this case the "autodetected" api address.

**Comment by mmack** [comment]
`There is a bug that if the token command fails to parse the config, it'll just silently revert to use default config, in this case the "autodetected" api address.`

I think that was my problem indeed! Thanks for the help!

**Comment by ruskofd** [comment]
Well from my side, I think I see why, it's because I don't specify the configuration when I generate a token :

```
- name: Generate controller join token
  ansible.builtin.command: k0s token create --role=controller --expiry=1h
  register: generated_controller_token

- name: "Generate worker join token"
  ansible.builtin.command: k0s token create --role=worker --expiry=1h
  register: generated_server_token
  ```
My configuration is not stored directly next to the binary but in `/etc`, that's probably the explaination. I will check as soon as I can, and report here. 

However in the docs, it's not clearly specified, probably there is few details to add here :

```
To be able to join workers into the cluster we need a token. The token embeds information with which we can enable mutual trust between the worker and controller(s) and allow the node to join the cluster as worker.

To get a token run the following on one of the existing controller nodes:

k0s token create --role=worker
```

I can open an issue if needed.

Regards

**Comment by jnummelin** [comment]
I opened #705 to make token, and all other commands too, to fail with error when config parsing fails.
